pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go

```diff
@@ -185,48 +221,38 @@ func (n *klusterletController) sync(ctx context.Context, controllerContext facto
 	// Start deploy klusterlet components
 	// Ensure the existence namespaces for klusterlet and klusterlet addon
 	// Sync pull secret to each namespace
-	namespaces := []string{config.KlusterletNamespace, fmt.Sprintf("%s-addon", config.KlusterletNamespace)}
-	for _, namespace := range namespaces {
-		err := n.ensureNamespace(ctx, klusterlet.Name, namespace)
-		if err != nil {
-			return err
-		}
-
-		// Sync pull secret
-		_, _, err = resourceapply.SyncSecret(
-			n.kubeClient.CoreV1(),
-			controllerContext.Recorder(),
-			n.operatorNamespace,
-			imagePullSecret,
-			namespace,
-			imagePullSecret,
-			[]metav1.OwnerReference{},
-		)
+	err = n.syncPullSecret(ctx, n.kubeClient, n.kubeClient, klusterlet.Name, config.KlusterletNamespace, controllerContext.Recorder())
+	if err != nil {
+		return err
+	}
+	// For now, whether in Default or Detached mode, the addons will be deployed on the managed cluster.
+	// sync image pull secret from management cluster to managed cluster for addon namespace
+	// TODO(zhujian7): In the future, we may consider deploy addons on the management cluster in Detached mode.
+	addonNamespace := fmt.Sprintf("%s-addon", config.KlusterletNamespace)
+	err = n.syncPullSecret(ctx, n.kubeClient, managedClusterClients.kubeClient, klusterlet.Name, addonNamespace, controllerContext.Recorder())
+	if err != nil {
+		return err
+	}
 
+	if detachedMode {
+		// In detached mode, we should ensure the namespace on the managed cluster since
+		// some resources(eg:role, rolebinding) are still deployed on managed cluster.
+		err := n.ensureNamespace(ctx, managedClusterClients.kubeClient, klusterletName, config.KlusterletNamespace)
 		if err != nil {
 			return err
 		}
 	}
 
-	if err != nil {
-		_, _, _ = helpers.UpdateKlusterletStatus(ctx, n.klusterletClient, klusterletName, helpers.UpdateKlusterletConditionFn(metav1.Condition{
-			Type: klusterletApplied, Status: metav1.ConditionFalse, Reason: "KlusterletApplyFailed",
-			Message: fmt.Sprintf("Failed to sync image pull secret to namespace %q: %v", config.KlusterletNamespace, err),
-		}))
-
-		return err
-	}
-
 	var relatedResources []operatorapiv1.RelatedResourceMeta
-	errs := []error{}
 	// If kube version is less than 1.12, deploy static resource for kube 1.11 at first
 	// TODO remove this when we do not support kube 1.11 any longer
 	if cnt, err := n.kubeVersion.Compare("v1.12.0"); err == nil && cnt < 0 {
+		errs := []error{}
 		resourceResult := resourceapply.ApplyDirectly(
-			resourceapply.NewKubeClientHolder(n.kubeClient),
+			resourceapply.NewKubeClientHolder(managedClusterClients.kubeClient),
 			controllerContext.Recorder(),
 			func(name string) ([]byte, error) {
-				template, err := manifests.Klusterlet111ManifestFiles.ReadFile(name)
+				template, err := manifests.KlusterletManifestFiles.ReadFile(name)
```

Comment:
why change this line?

Reply:
I put all klusterlet files in `KlusterletManifestFiles` to avoid duplication of code